connection plugin kubectl for kubernetes.

xuxinkun opened the pull request:
##### SUMMARY

This PR is the connection plugin for kubernetes. It can deploy the playbook directly into containers of kubernetes pod using the local `kubectl`.

kubernetes has a tool to enter container called `kubectl`. It has similar feature as `docker exec`.
So, by making use of `kubectl`, we can alse achieve the goal treating the `kubernetes pod` as container of docker.

After this PR, inventory example could be like this.

```
[k8s-pod]
mypod ansible_connection=kubernetes ansible_kubernetes_extra_args="-s http://101.88.64.106:8080" ansible_kubernetes_namespace=xuxinkun ansible_kubernetes_container=master
```

The following parameters are processed by this connector:

- ansible_kubernetes_container: If the pod has more than one container, the container name must be specified.
- ansible_kubernetes_namespace: The namespace of pod. If it is not set, it will be `default`.
- ansible_kubernetes_extra_args: Could be a string with any additional arguments understood by kubectl, which are not command specific. This parameter is mainly used to configure the options of `kubectl`.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME

- contrib/inventory/kubernetes.py
- lib/ansible/plugins/connection/kubernetes.py

##### ANSIBLE VERSION

```
devel
```


##### ADDITIONAL INFORMATION

I also see that there is a similar PR https://github.com/ansible/ansible/pull/24960. But I do not agree with it. First, kubernetes version has changed almost every two months. So, using python library `kubernetes` can not be compatible with each version. And it did not implement `fetch_file`.

I take use of `kubectl` instead of python library `kubernetes`. 

```
[root@ab22a1cf08be ~]# kubectl help exec
Execute a command in a container.

Examples:
  # Get output from running 'date' from pod 123456-7890, using the first container by default
  kubectl exec 123456-7890 date
  
  # Get output from running 'date' in ruby-container from pod 123456-7890
  kubectl exec 123456-7890 -c ruby-container date
  
  # Switch to raw terminal mode, sends stdin to 'bash' in ruby-container from pod 123456-7890
  # and sends stdout/stderr from 'bash' back to the client
  kubectl exec 123456-7890 -c ruby-container -i -t -- bash -il

Options:
  -c, --container='': Container name. If omitted, the first container in the pod will be chosen
  -p, --pod='': Pod name
  -i, --stdin=false: Pass stdin to the container
  -t, --tty=false: Stdin is a TTY

Usage:
  kubectl exec POD [-c CONTAINER] -- COMMAND [args...] [options]
```

And this has not been changed and will not be changed for a long period of time. 

But the options has been changed a lot. So, I set `ansible_kubernetes_extra_args` to replace it. Ansible user can set it easily.

xuxinkun committed:
Adds kubectl and oc connection plugins

xuxinkun commented:
@bcoca Pls review and check. Kubernetes is more and more popular. I hope this could be merged to help kubernetes users.

Ledger45 commented:
Replacing python api with kubectl is a good idea

chrrrles reviewed (approved):
Did a cursory connection test using the raw module and several Openshift containers. Worked good for that.

chrrrles commented on lib/ansible/plugins/connection/kubernetes.py:
```diff
@@ -0,0 +1,179 @@
+# Based on the kubernetes connection plugin
```
Is this based on JCPowermac's connection plugin?

xuxinkun replied:
based on docker connection. I missed it. I will fix it right now.

m8cool commented:
Wonderful feature. I cannot wait to use this plugin.

chouseknecht commented:
@xuxinkun @chrrrles 

Did some testing with OpenShift as well. I could only get it to work with the `raw` module. Why is that?

To get it to work I created a soft link for kubectl -> oc, as I don't have kubectl installed. Then I created the following inventory file:

```
django-3-6jt7z ansible_kubernetes_namespace=demo
```
In the above `django-3-6jt7z` is the name of the pod. It only contains a single command. And, `demo` is the name of the project.

Again, the question is, why does the `raw` module work, but `command` does not? And a second question, why does `gather_facts` not work?

In both cases (i.e. using `command`, or enabling `gather_facts`), I get the following error:

```
UNREACHABLE! => {"changed": false, "msg": "Authentication or permission failure. In some cases, you may have been able to authenticate and did not have permissions on the target directory. Consider changing the remote temp path in ansible.cfg to a path rooted in \"/tmp\". Failed command was: ( umask 77 && mkdir -p \"` echo /.ansible/tmp/ansible-tmp-1499887864.83-262240610026274 `\" && echo ansible-tmp-1499887864.83-262240610026274=\"` echo /.ansible/tmp/ansible-tmp-1499887864.83-262240610026274 `\" ), exited with result 1", "unreachable": true
```

FWIW, here's my playbook using `raw`:

```
- name: Talk to containers on OpenShift
  hosts: django-3-6jt7z
  gather_facts: no
  connection: kubernetes
  tasks:

    - raw: date
      register: output

    - debug: var=output.stdout
```

UPDATE: 

I think the issue is likely that the container is running as an unknown user. That's the default OpenShift behavior. It runs your container as a random user. So perhaps the only thing that will ever work in a default OpenShift installation is the `raw` command.

```
$ oc exec django-3-6jt7z whoami
whoami: cannot find name for user ID 1000060000
```

chrrrles commented:
UPDATE:
Chris - you are correct about the user id -- The "pet" pod below is deployed with a root service account, so is able to run modules.


Hi Chris,

I am testing on my local minishift, using a `kubectl` installed by `brew`.  Try running the playbook on containers with a mounted, writable volume ...  I believe that is required in order for this connection to work with modules other than `raw`.   Here is a playbook and a run on a "pet" pod with writable volumes.

```
- hosts: mypod
  tasks:
          - name: echo "Hello"
            raw: echo "HELLO raw"
            register: output
          - debug: var=output.stdout
          - name: command echo "Hello"
            command: echo "HELLO command"
            register: output
          - debug: var=output.stdout
          - name: command echo "Hello"
            shell: echo "HELLO shell"
            register: output
          - debug: var=output.stdout
```

```
PLAY [mypod] *****************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************************************************************************************************************************************************
ok: [aerospike-2-4md2v]

TASK [echo "Hello"] **********************************************************************************************************************************************************************************************************************************************
changed: [aerospike-2-4md2v]

TASK [debug] *****************************************************************************************************************************************************************************************************************************************************
ok: [aerospike-2-4md2v] => {
    "failed": false,
    "output.stdout": "HELLO raw\n"
}

TASK [command echo "Hello"] **************************************************************************************************************************************************************************************************************************************
changed: [aerospike-2-4md2v]

TASK [debug] *****************************************************************************************************************************************************************************************************************************************************
ok: [aerospike-2-4md2v] => {
    "failed": false,
    "output.stdout": "HELLO command"
}

TASK [command echo "Hello"] **************************************************************************************************************************************************************************************************************************************
changed: [aerospike-2-4md2v]

TASK [debug] *****************************************************************************************************************************************************************************************************************************************************
ok: [aerospike-2-4md2v] => {
    "failed": false,
    "output.stdout": "HELLO shell"
}

PLAY RECAP *******************************************************************************************************************************************************************************************************************************************************
aerospike-2-4md2v          : ok=7    changed=3    unreachable=0    failed=0

xuxinkun commented:
@chouseknecht  I don't know about how `oc` works with exec. But this pulgin work well with origin `kubectl`.
Here is my inventory:
```
[k8s]
centos-pod ansible_kubernetes_extra_args="-s http://10.8.64.106:8080" ansible_kubernetes_namespace=xuxinkun
```

and my playbook:
```
- name: Talk to containers on OpenShift
  hosts: k8s
  connection: kubernetes
  tasks:
    - raw: date
      register: output
    - debug: var=output.stdout
    - command: date
      register: cmd_output
    - debug: var=cmd_output.stdout
```
And here is my result:

```
[root@ab22a1cf08be ~]# ansible-playbook playbook  -v
Using /etc/ansible/ansible.cfg as config file

PLAY [Talk to containers on OpenShift] ******************************************************************************

TASK [Gathering Facts] **********************************************************************************************
ok: [centos-pod]

TASK [raw] **********************************************************************************************************
changed: [centos-pod] => {"changed": true, "failed": false, "rc": 0, "stderr": "", "stdout": "Thu Jul 13 01:52:35 UTC

TASK [debug] ********************************************************************************************************
ok: [centos-pod] => {
    "failed": false, 
    "output.stdout": "Thu Jul 13 01:52:35 UTC 2017\n"
}

TASK [command] ******************************************************************************************************
changed: [centos-pod] => {"changed": true, "cmd": ["date"], "delta": "0:00:01.081603", "end": "2017-07-13 01:52:37.81_lines": [], "stdout": "Thu Jul 13 01:52:36 UTC 2017", "stdout_lines": ["Thu Jul 13 01:52:36 UTC 2017"]}

TASK [debug] ********************************************************************************************************
ok: [centos-pod] => {
    "cmd_output.stdout": "Thu Jul 13 01:52:36 UTC 2017", 
    "failed": false
}

PLAY RECAP **********************************************************************************************************
centos-pod                 : ok=5    changed=2    unreachable=0    failed=0 
```

xuxinkun commented:
@chouseknecht  BTW, as I know, the `kubectl exec` could not specify the remote user right now. But docker client has achieved that. Maybe kubectl will achieve it in the future. I'm not sure. So my connection plugin can not support specifying `remote user`. Users can only use the `user` which container started with.
And also, some images or containers does not have `python` or `dd`. And this plugin can not work with them too.

jcpowermac commented:
@chouseknecht take a look at my [test project](https://github.com/jcpowermac/ansible-kube-connect-test/blob/master/python/Dockerfile) for my iteration for the kube connection plugin.  I used an example from [my team's repo](https://github.com/RHsyseng/container-rhel-examples) that deals with arbitrary uid.  Arbitrary uid just needs to be handled in the Docker image.  I would figure this implementation would have no problem with any modules as long as the uid issue was handled.

xuxinkun commented:
@bcoca After this PR is merged, I will make another PR for docs of kubectl connection plugin.

xuxinkun commented:
Is there any progress about this PR? @bcoca @jcpowermac

jcpowermac commented:
@bcoca @chrrrles what is needed for this PR to be merged?
After this is merged I can continue my PR.

bcoca commented:
im trying to get ansible config merged first as it changes how plugin configurations work (no more need to modify play_context) .. both these plugins can still make it in after tomorrow's freeze

jcpowermac commented:
awesome thanks @bcoca !

xuxinkun commented:
@bcoca  It seemed that https://github.com/ansible/ansible/pull/27448 is not a completed PR. 
For example, https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/play_context.py#L318 
 ```
setattr(self, flag, self.connection.get_option(flag))
```
`self.connection` is a string, and does not have method `get_option`. Maybe It should be `plugin.get_option`? 
I am confused to set the option for playcontext. Pls help.

bcoca commented:
@xuxinkun that is for backwards compatibilty, this is the future plan #28473

xuxinkun commented:
@bcoca  So, I still need to merge the option such as ` kube_config_file/kube_container/kube_namespace` to  `play_context.py`. Is it right?

bcoca commented:
do that for now, i'll move it to the new options once i get that merged

xuxinkun commented:
@bcoca  OK. I have rebased my PR and pushed it again. Pls check.

xuxinkun commented:
@bcoca @jcpowermac I have rebased my PR. Is there anything needed to change before merge?

xuxinkun commented:
@jcpowermac  @bcoca Is there any progress about this PR? or something needed to modify?

bcoca commented:
yes, sadly it took me too long to revamp the config, so it will be merged into 2.5 now, i'll update the PR with the new config (which i should be merging this week or early next).

xuxinkun commented:
@bcoca  OK. I will continue to focus this PR. If there is more thing to be modified, just ping me.

SpamapS commented:
This looks super promising. Can we pile on and get this merged? Looks like it's conflicting right now.

xuxinkun commented:
@Spamaps bcoba and some other members are working on the config. So, my PR will be delayed until they finish it.

bcoca commented:
just finished last friday, @xuxinkun if you can look at current connection pugins and use the 'documentation' instead of updating play context, we can get this merged soon

xuxinkun commented:
@bcoca  OK. I will modify my PR as soon as possible.

xuxinkun commented:
@bcoca  I have finished it. I have read the code and use `get_option` instead of `_play_context`.  Am I right? Pls review.

xuxinkun commented:
@bcoca I've finished it. Have you ever reviewed this PR?

xuxinkun commented:
ping @maxamillion

maxamillion commented on lib/ansible/plugins/connection/kubectl.py:
```diff
@@ -0,0 +1,220 @@
+    short_description: Run tasks in kubernetes pods
+    description:
+        - Run commands or put/fetch files to an existing kubernetes pod container.
+    version_added: "2.0"
```
Bump this to `"2.5"` please.

chouseknecht commented:
@xuxinkun 

I have a few issues with this PR...

Firstly, we've been moving toward the Python client (see the new things we've pushed into 2.5). We can debate the pros/cons of the Python client, but for now, it's the direction we've chosen. We *should* be able to replicate this PR using the Python client, and so if we decide that a connection plugin is necessary, I'd like us to move in that direction.

Secondly, we should consider the virtues of initiating Ansible playbooks, or any other commands, from outside of the platform to then run inside of an *immutable* container. Now that Jobs are fully functional, we should be using those to initiate playbooks and other actions.
 
Thirdly, we should bounce this off of the K8s Ansible Config WG, chaired by @smarterclayton There is a Google mailing list here: https://groups.google.com/forum/#!forum/kubernetes-ansible-config-wg, which I'm assuming has public read access.

Fourthly, introducing this potentially leads to another round of bastardization of the `synchronize` plugin, and I think everyone in the known universe is, or should be, opposed to such. All container connection plugins lead down that road, at least they have thus far.

maxamillion commented:
> 
> Firstly, we've been moving toward the Python client (see the new things we've pushed into 2.5). We can debate the pros/cons of the Python client, but for now, it's the direction we've chosen. We *should* be able to replicate this PR using the Python client, and so if we decide that a connection plugin is necessary, I'd like us to move in that direction.
> 

+1 - I did want to bring this up for discussion

> Secondly, we should consider the virtues of initiating Ansible playbooks, or any other commands, from outside of the platform to then run inside of an *immutable* container. Now that Jobs are fully functional, we should be using those to initiate playbooks and other actions.
>  

I'm on the fence with this one, I don't personally have an use case for using this connection plugin but I'm open to the reality that others might need it.

> Thirdly, we should bounce this off of the K8s Ansible Config WG, chaired by @smarterclayton There is a Google mailing list here: https://groups.google.com/forum/#!forum/kubernetes-ansible-config-wg, which has public read access.
> 

+1 - would definitely like to hear the thoughts of the K8s Ansible Config WG and @smarterclayton in particular

> Fourthly, introducing this potentially leads to another round of bastardization of the `synchronize` plugin, and I think everyone in the known universe is, or should be, opposed to such. All container connection plugins lead down that road, at least they have thus far.
>

I don't have enough background information on all the other container connection plugins to have an informed opinion here, so I'll defer to those who do.

chrrrles commented:
@chouseknecht  When will a `Jobs` based alternative to running (or not running...) ad-hoc container commands be ready?

Also, that WG link is private, and not for plebes like us to view:
```
 kubernetes-ansible-config-wg
You must be a member of this group to view and participate in it. Membership is by invitation only.

This group is invite only, so you may not apply for membership.
```

Not going to get involved in the python client debate, it's not my war to fight.  That said, how do we get involved in this community?  Or is it even worth it?

chouseknecht commented:
@chrrrles 

Maybe the Google group is invitees only. Sorry. Thought it had public read access. 

At any rate, I'll make sure folks in that group are aware of this PR, and get them to comment.

You can use the `k8s_raw` and `openshift_raw` modules now in 2.5 to launch jobs.

chouseknecht commented:
Thinking about a connection plugin in the context of inventory, I'm wondering if inventory is about workloads running on the platform, or the platforms themselves. You might say it's about both, but I'm not sure that's right. Imagine the typical large scale shop that runs multiple clusters.

In the VM world, the inventory is a farm of VMs. We group them according to subject or context, and then associate connection details with a group, or perhaps at the individual VM level. But what we don't put in inventory is the workload running on each VM. We don't inventory all the individual processes.

The point of inventory is to provide details for context switching. In the case of VMs it's about switching the context from one VM to the next, and running the appropriate tasks, with the right options in the current context.

Now apply that concept to container platforms. What's the context? I don't think it should be containers or workloads. The context is the platform, and the connection details are about the platform. What we need to do at the Ansible play level, is set the context to the correct cluster, connect using the appropriate service account (or token, certs, etc.), and switch to the right namespace (or project). Once in the right context, then fire the tasks. 

A connection plugin for K8s should be about context switching with regard to platform, and not about workloads or containers, and our inventory file should be a reflection of the .kube/config file.

chouseknecht commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Connection plugin](https://trello.com/c/pFWVUuA4/16-connection-plugin)

chouseknecht commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Connection plugin](https://trello.com/c/pFWVUuA4/16-connection-plugin)

chrrrles commented:
Oh cool.  Thank you @chouseknecht

chrrrles commented:
@chouseknecht - If an inventory file only references clusters (or contexts, rather) then how would we select the container/pod to use at the play level...  Would this be done by adding new parameters to Playbook?

Also, in terms of 'portability' of existing playbooks - I feel it would be easier to just redefine the existing inventory item as a k8s connection to a specific pod/deployment/container rather than modifying the playbooks with selectors for k8s resources.

Will the 'setup facts' feature from the Trello board have a role in all this?

chouseknecht commented:
@chrrrles 

From the discussion in the Kubernetes Ansible Config WG, it's pretty clear that the connection plugin must connect to individual containers, and support 'exec' commands.

Here's some of the use cases the group enumerated:

1. I want to connect to one or more clusters to install a component into each cluster that is a singleton - things like the openshift registry, or router, or api extensions, or operators, etc, and then I will later on run playbooks that upgrade those components
2. I want to manage one or more “applications” (defined as loosely as possible, ie not as a concrete object but a loose collection of resources) on a cluster, but I might have multiple of these in different roles (dev, stage, prod, or regional, or for different teams)
3. I want to manage one or more “applications” across many clusters - blending the use case above but with the distinction that I might upgrade parts of the application in order
4. I want to manage a large number of similar applications on one or more clusters, and I want to dynamically populate that inventory via an api.  This is the use case for Ansible service broker and ansible playbook having its own api to track instances that can give an inventory (ie the service broker already tracked these, so we want to build inventory from that api to run maintenance / migration)

This PR does that, but I'd still like to see if we can do the following before landing on a final solution:

- Move to using the OpenShift Python client
- Explore whether or not a connection plugin can handle both the API level connection, and optional the 'exec' connection to specific pod/container. Don't know if it's possible, but would be super cool, if it could accommodate both.

jcpowermac commented:
@chouseknecht I already have a [PR](https://github.com/ansible/ansible/pull/24960) opened using the python client.  I also have an inventory plugin developed but I need the existing PR merged before submitting the inventory plugin.

jcpowermac commented:
Correction it uses: https://github.com/kubernetes-incubator/client-python
This fell off my radar since it was waiting for 2.5 based changes.

chouseknecht commented:
@jcpowermac 

Awesome! Didn't know that was out there. Will take a look shortly.

chouseknecht commented:
After more discussions and deliberations, it seems the better path forward is to wrap the `kubectl` binary, similar to how we've done other connection plugins, so +1 for this PR.

My only ask then, can we bring the available options and names for the connection plugin inline with the modules. The modules currently support the following authorization options, which I think the connection plugin should also accept:

- username
- password
- api_key
- host
- kubeconfig
- verify_ssl
- context
- key_file
- cert_file
- ssl_ca_cert

chrrrles commented:
@xuxinkun - have you followed the discussion?

xuxinkun commented:
@chouseknecht Have these options already been discussed? It is not hard to accept these. But it will take some days for me to modify and test.
@chrrrles I will try to modify the PR as soon as possible.

chouseknecht commented:
@xuxinkun 

Will work on pushing these changes into your branch today.

As far as discussion around parameter specifics, totally open to a discussion. If anyone has ideas, or wants to make changes, now's the time. I've simply attempted to surface the Python client parameters, keeping the names and meaning consistent with the client. But again, open to modifying this approach, and iterating on it to make it better.

jstoja commented:
Maybe add the `namespace` as a connection parameter?

chouseknecht commented:
@jstoja 

Good idea. Namespace is definitely included.

Added a bunch of variables, and support for env variables too. I think parameters are now pretty well aligned with the modules.

xuxinkun commented on lib/ansible/plugins/connection/kubectl.py:
```diff
@@ -23,47 +23,127 @@
+          - name: ansible_kubectl_cert_file
+        env:
+          - name: K8S_AUTH_SSL_CA_CERT
+      verify_ssl:
```
How about using `kubectl_verify_ssl` instead of `verify_ssl`?

chouseknecht replied:
Yep. That's a bug. Fixing.

xuxinkun commented on lib/ansible/plugins/connection/kubectl.py:
```diff
@@ -23,47 +23,127 @@
+          - name: ansible_kubectl_context
+        env:
+          - name: k8S_AUTH_CONTEXT
+      kubectl_host:
```
I prefer to use `server` instead of `host`. Because `host` in ansible points to the target of operation.
 BTW, kubectl also use `-s, --server string                                The address and port of the Kubernetes API server`

chouseknecht replied:
I know. There's an alias for 'server'. Calling it `host` keeps it consistent with the modules and Python client. As a user, you can use `server` or `host`. The Python client calls it `host`, and uses `HOST` in the env variable.

xuxinkun replied:
How about adding a `SERVER` env variable?

chouseknecht replied:
Done.

xuxinkun commented on lib/ansible/plugins/connection/kubectl.py:
```diff
@@ -23,47 +23,127 @@
         vars:
-            - name: ansible_kube_namespace
-            - name: ansible_kubectl_namespace
+          - name: ansible_kubectl_container
```
Should we also provide a env option for `kubectl_container`?

chouseknecht replied:
Doesn't seem like it would be necessary, as container names aren't really known ahead of time. I'm envisioning pod and container names coming from an inventory script.

xuxinkun replied:
If pod has only one container, container name can be ignored. But if pod has more than one container, container name is necessary to be specified.

chouseknecht replied:
Added.

chouseknecht commented:
Added a `oc` connection plugin that subclasses the kubectl plugin. That should suffice if you only have access to the `oc` binary, or just want to test it directly for some reasons. It has slightly different options, plus it has an `rsync` command, which may prove useful.

xuxinkun commented on lib/ansible/plugins/connection/kubectl.py:
```diff
@@ -23,47 +23,127 @@
 DOCUMENTATION = """
     author:
         - xuxinkun
+    
```
Remove these blank lines.
